### PR TITLE
Implement Head Yaw.

### PR DESCRIPTION
**Pull request by TheBlackPlague**
## Introduction

Currently, PM doesn't really consider head-yaw to be a property. This can be tedious/nearly impossible when you want to change head-yaw without messing directly with the protocol. 

### Relevant issues

- There isn't really an issue for this.

### API changes

Location now supports a property `headYaw`. The property is strictly of type `float`. The constructor accepts this as an optional argument. All methods used to change the rotation of an entity also use the `headYaw` property if set. Furthermore, all methods mentioned above accept this as an optional argument. 

### Behavioural changes

The server should continue to behave the same way. If no `headYaw` is provided, then by default, the player's `yaw` is substituted in.

## Backwards compatibility

To preserve BC, the argument has been added in as optional to all methods. Meaning plugins or internal functions not supplying a `headYaw` argument to either the constructor or changed methods will continue to work the same way with no behavior change. If they provide the optional field, they'll be able to use it. 

A possible issue is the methods extending the `Location` class. They might need to update the relevant signature for their constructor. A possible fix for this issue is by removing it from a constructor and adding it as a function-based setter.

## Tests

This adds a property that doesn't break any internal BC. Henceforth, it doesn't necessarily need to be tested as it behaves the same way.

### CODE REVIEW

File: src/player/Player.php [hunk 1]
```diff
@@ -2341,7 +2341,7 @@ public function broadcastSound(Sound $sound, ?array $targets = null) : void{
 	/**
 	 * TODO: remove this
 	 */
-	protected function sendPosition(Vector3 $pos, ?float $yaw = null, ?float $pitch = null, int $mode = MovePlayerPacket::MODE_NORMAL) : void{
+	protected function sendPosition(Vector3 $pos, ?float $yaw = null, ?float $pitch = null, int $mode = MovePlayerPacket::MODE_NORMAL, ?float $headYaw = null) : void{
 		$this->getNetworkSession()->syncMovement($pos, $yaw, $pitch, $mode);
```
Comment (JavierLeon9966) [inline src/player/Player.php:2345]:
```suggestion
		$this->getNetworkSession()->syncMovement($pos, $yaw, $pitch, $mode, $headYaw);
```

File: src/player/Player.php [hunk 2]
```diff
@@ -2350,7 +2350,7 @@ protected function sendPosition(Vector3 $pos, ?float $yaw = null, ?float $pitch
 	/**
 	 * {@inheritdoc}
 	 */
-	public function teleport(Vector3 $pos, ?float $yaw = null, ?float $pitch = null) : bool{
+	public function teleport(Vector3 $pos, ?float $yaw = null, ?float $pitch = null, ?float $headYaw = null) : bool{
 		if(parent::teleport($pos, $yaw, $pitch)){
```
Comment (JavierLeon9966) [inline src/player/Player.php:2354]:
```suggestion
		if(parent::teleport($pos, $yaw, $pitch, $headYaw)){
```

**Comment by dktapps** [comment]
I'm doubtful this should be in Location, since Location is generic for all entities, but objects such as arrows don't have head yaw.

**Review by JavierLeon9966 — CHANGES_REQUESTED** [review]
You forgot to pass `$headYaw` in some functions so here’s a list including this one: https://github.com/pmmp/PocketMine-MP/blob/00717e15d4762ca93ad4a991062b343e2e75a14d/src/player/Player.php#L2358

**Comment by TheBlackPlague** [comment]
> I'm doubtful this should be in Location, since Location is generic for all entities, but objects such as arrows don't have head yaw.

Where do you suggest it should be then? In Entity, I presume then? Or in the Human?

**Comment by dktapps** [comment]
All entities descend from Entity, that's no better than Location.

I would suggest Living, but it's kinda awkwardly placed in there and there's a lot more needs to be done to properly support it (e.g. lookAt() needs to be aware of it).

**Comment by TheBlackPlague** [comment]
> All entities descend from Entity, that's no better than Location.
> 
> I would suggest Living, but it's kinda awkwardly placed in there and there's a lot more needs to be done to properly support it (e.g. lookAt() needs to be aware of it).

I can implement such in Living, that shouldn't be an issue, along with making `lookAt()` aware of it. Would you be able to give me a list of methods that use the yaw in a way that headYaw would be a better use? That would speed up my development time a lot.

**Comment by JavierLeon9966** [comment]
Also `getDirectionVector()`,`getHorizontalFacing()`, `sendSpawnPacket()`, `setPosition()`, `setRotation()`, `setPositionAndRotation()`, ` broadcastMovement()`, `move()`

**Comment by TheBlackPlague** [comment]
Furthermore, I am also unsure why all the variables in Location/Vector3 remain public— there should be Getters and Setters for those properties. Ensuring BC becomes 10x harder if properties are publicized unnecessarily. 

A way to counter this is using the `Magic Methods` (`__set()` and `__get()`). Your thoughts @dktapps ?

**Comment by dktapps** [comment]
I think it's pointless. This whole yarn about BC is nonsense. __get() and __set() exist exactly for the purpose of avoiding BC breaks. Adding them at this point serves no advantage. In addition, the fields of Location and Vector3 should not be mutable, so we don't want setters anyway. But that's outside the scope of this PR.

**Comment by dktapps** [comment]
I think it's also worth mentioning that one of the reasons why Vector3 in particular continues to have public members is because it's used in hot pathways in the core code. I don't like having them public (or, more accurately, I don't like having them _mutable_), but it's currently a tradeoff that I feel ok about making.

**Comment by TheBlackPlague** [comment]
After further discussion with @dktapps , it was decided that this cannot be reasonably implemented in a way that's good for the code-base & is backward compatible. What this means is that there is a good chance this will be in `5.0.0` (the next major release), but it will not be in any `4.x.x` (minor releases).

File: src/entity/Entity.php [hunk 3]
```diff
@@ -1331,15 +1331,24 @@ protected function setPosition(Vector3 $pos) : bool{
...
 		$this->scheduleUpdate();
 	}
 
-	protected function setPositionAndRotation(Vector3 $pos, float $yaw, float $pitch) : bool{
+	protected function setPositionAndRotation(Vector3 $pos, float $yaw, float $pitch, float $headYaw = -1) : bool{
 		if($this->setPosition($pos)){
-			$this->setRotation($yaw, $pitch);
+			if ($headYaw < 0) {
+				$headYaw = $yaw;
+			}
```
Comment (ShockedPlot7560) [inline src/entity/Entity.php:1349]: Already done in setRotation, no need to write again

File: src/entity/Entity.php [hunk 4]
```diff
@@ -1388,9 +1397,12 @@ public function isOnGround() : bool{
 	/**
 	 * @param Vector3|Position|Location $pos
 	 */
-	public function teleport(Vector3 $pos, ?float $yaw = null, ?float $pitch = null) : bool{
+	public function teleport(Vector3 $pos, ?float $yaw = null, ?float $pitch = null, ?float $headYaw = null) : bool{
 		if($pos instanceof Location){
 			$yaw = $yaw ?? $pos->yaw;
+			/** @noinspection CallableParameterUseCaseInTypeContextInspection - This will never happen. */
```
Comment (ShockedPlot7560) [inline src/entity/Entity.php:1403]: this noinspection comment has no place to be here

File: src/entity/Location.php [hunk 5]
```diff
@@ -32,32 +32,45 @@ class Location extends Position{
 	/** @var float */
 	public $yaw;
 	/** @var float */
+	public $headYaw;
```
Comment (ShockedPlot7560) [inline src/entity/Location.php:35]: HeadYaw has no place to be in Location.

**Comment by ShockedPlot7560** [comment]
Need to target next-minor

**Comment by ShockedPlot7560** [comment]
Oops, missclick on close

**Comment by dktapps** [comment]
The PR is very stale, might as well close it and open a tracking issue instead

**Pull request closed without merging**